src/ray/common/nullable_uint.h

```diff
@@ -0,0 +1,28 @@
+
+#pragma once
+
+#include <stdint.h>
+
+#include <optional>
+
+namespace ray {
+
+/// Represents a nullable uint64_t using int64_t, using the most significant bit to
+/// represent null - when the bit is set the value is null.
+typedef int64_t nuint64_t;
+
+/// Represents null for a nullable integer value
+static constexpr int64_t kNull = -1;
+
+/// Defines a nullable, unsigned integer.
+class NullableInt {
```

Comment:
We can just use std::optional I think since we are using c++17.

Reply:
i thought about using optional but it is not as clean, e.g.

```
nuint64_t NullableInt::min(nuint64_t left, nuint64_t right) {
  NullableInt::assertValidNullableInt(left);
  NullableInt::assertValidNullableInt(right);

  if (left == kNull && right == kNull) {
    return kNull;
  } else if (left == kNull) {
    return right;
  } else if (right == kNull) {
    return left;
  } else {
    return std::min(left, right);
  }
}
```

```
nuint64_t NullableInt::min(nuint64_t left, nuint64_t right) {
  if (!left && !right) {
    return std::nullopt;
  } else if (left == kNull) {
    return right;
  } else if (right == kNull) {
    return left;
  } else {
    return std::optional(std::min(*left, *right));
  }
} 
```

Reply:
Example 2:

```
std::tuple<nuint64_t, nuint64_t> MemoryMonitor::GetCGroupMemoryBytes() {
  nuint64_t total_bytes = kNull;
  if (std::filesystem::exists(kCgroupsV2MemoryMaxPath)) {
    std::ifstream mem_file(kCgroupsV2MemoryMaxPath, std::ios::in | std::ios::binary);
    mem_file >> total_bytes;
  } else if (std::filesystem::exists(kCgroupsV1MemoryMaxPath)) {
    std::ifstream mem_file(kCgroupsV1MemoryMaxPath, std::ios::in | std::ios::binary);
    mem_file >> total_bytes;
  }

  nuint64_t used_bytes = kNull;
  if (std::filesystem::exists(kCgroupsV2MemoryUsagePath)) {
    std::ifstream mem_file(kCgroupsV2MemoryUsagePath, std::ios::in | std::ios::binary);
    mem_file >> used_bytes;
  } else if (std::filesystem::exists(kCgroupsV1MemoryUsagePath)) {
    std::ifstream mem_file(kCgroupsV1MemoryUsagePath, std::ios::in | std::ios::binary);
    mem_file >> used_bytes;
  }

  RAY_CHECK((total_bytes == kNull && used_bytes == kNull) ||
            (total_bytes != kNull && used_bytes != kNull));
  if (total_bytes != kNull) {
    RAY_CHECK_GT(used_bytes, 0);
    RAY_CHECK_GT(total_bytes, used_bytes);
  }

  return {used_bytes, total_bytes};
}
```

```
std::tuple<nuint64_t, nuint64_t> MemoryMonitor::GetCGroupMemoryBytes() {
  nuint64_t total_bytes;
  if (std::filesystem::exists(kCgroupsV2MemoryMaxPath)) {
    std::ifstream mem_file(kCgroupsV2MemoryMaxPath, std::ios::in | std::ios::binary);
    mem_file >> *total_bytes;
  } else if (std::filesystem::exists(kCgroupsV1MemoryMaxPath)) {
    std::ifstream mem_file(kCgroupsV1MemoryMaxPath, std::ios::in | std::ios::binary);
    mem_file >> *total_bytes;
  }

  nuint64_t used_bytes = kNull;
  if (std::filesystem::exists(kCgroupsV2MemoryUsagePath)) {
    std::ifstream mem_file(kCgroupsV2MemoryUsagePath, std::ios::in | std::ios::binary);
    mem_file >> *used_bytes;
  } else if (std::filesystem::exists(kCgroupsV1MemoryUsagePath)) {
    std::ifstream mem_file(kCgroupsV1MemoryUsagePath, std::ios::in | std::ios::binary);
    mem_file >> *used_bytes;
  }

  RAY_CHECK((total_bytes && used_bytes) ||
            (!total_bytes && !used_bytes));
  if (!total_bytes) {
    RAY_CHECK_GT(*used_bytes, 0);
    RAY_CHECK_GT(*total_bytes, *used_bytes);
  }

  return {used_bytes, total_bytes};
}
```

Reply:
Another optional is to start adding overloads for optionals to our core library functions

Reply:
cc @rkooo567

Reply:
I vote for std::optional, it's the standard way of handling nullable data.

Reply:
nullable_uint might be a bit overkill, how about just use -1 for invalid result in memory monitoring, and use signed integer?

Reply:
Hmm nullable_uint is just a typedef around a signed int int64_t,  we could just call is int64_t if the concern is around introducing a new "type"